go/libraries/doltcore/sqle/kvexec/merge_join.go

```diff
@@ -0,0 +1,498 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kvexec
+
+import (
+	"errors"
+	"io"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+func newMergeKvIter(
+	leftIter, rightIter prolly.MapIter,
+	joiner *prollyToSqlJoiner,
+	lrComparer, llComparer func(val.Tuple, val.Tuple, val.Tuple, val.Tuple) int,
+	leftNorm, rightNorm coveringNormalizer,
+	leftFilter, rightFilter sql.Expression,
+	joinFilters []sql.Expression,
+	isLeftJoin bool,
+	excludeNulls bool,
+) (*mergeJoinKvIter, error) {
+	return &mergeJoinKvIter{
+		leftIter:     leftIter,
+		rightIter:    rightIter,
+		joiner:       joiner,
+		lrCmp:        lrComparer,
+		llCmp:        llComparer,
+		leftNorm:     leftNorm,
+		rightNorm:    rightNorm,
+		joinFilters:  joinFilters,
+		leftFilter:   leftFilter,
+		rightFilter:  rightFilter,
+		isLeftJoin:   isLeftJoin,
+		excludeNulls: excludeNulls,
+	}, nil
+}
+
+type mergeJoinKvIter struct {
+	leftIter prolly.MapIter
+	leftKey  val.Tuple
+	leftVal  val.Tuple
+
+	rightIter prolly.MapIter
+	rightKey  val.Tuple
+	rightVal  val.Tuple
+
+	lrCmp func(val.Tuple, val.Tuple, val.Tuple, val.Tuple) int
+	llCmp func(val.Tuple, val.Tuple, val.Tuple, val.Tuple) int
+
+	// Non-covering secondary index reads are faster separated
+	// from the initial mapIter, because we avoid unnecessary
+	// primary reads. Note: keyless indexes are less amenable to
+	// this optimization because their cardinality is stored in the
+	// primary index.
+	leftNorm  coveringNormalizer
+	rightNorm coveringNormalizer
+
+	nextRightKey val.Tuple
+	nextRightVal val.Tuple
+
+	lookaheadBuf [][]byte
+	matchPos     int
+
+	// projections
+	joiner *prollyToSqlJoiner
+
+	// todo: we want to build KV-side static expression implementations
+	// so that we can execute filters more efficiently
+	joinFilters []sql.Expression
+	leftFilter  sql.Expression
+	rightFilter sql.Expression
+
+	// LEFT_JOIN impl details
+	excludeNulls bool
+	isLeftJoin   bool
+	matchedLeft  bool
+	exhaustLeft  bool
+}
+
+var _ sql.RowIter = (*mergeJoinKvIter)(nil)
+
+func (l *mergeJoinKvIter) Close(_ *sql.Context) error {
+	return nil
+}
+
+func (l *mergeJoinKvIter) Next(ctx *sql.Context) (sql.Row, error) {
+	var err error
+	if l.leftKey == nil {
+		if err := l.initialize(ctx); err != nil {
+			if errors.Is(err, io.EOF) && l.isLeftJoin {
+				goto exhaustLeft
+			}
+			return nil, err
+		}
+	}
+
+	if l.exhaustLeft {
+		goto exhaustLeft
+	}
+
+	if len(l.lookaheadBuf) > 0 || l.matchPos > 0 {
+		goto match
+	}
+
+incr:
```

Comment:
Maybe just call this `compare`

Reply:
The only exits from this block are to the next block (`matchBuf`) or to `exhaustLeft`, otherwise it loops.

`exhaustLeft` is terminal and is better expressed as a method (see below), so this whole block would be better expressed as a `for {` loop with a `break` statement when it's done